cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/AbstractKafkaSampleStore.java

```diff
@@ -63,23 +63,44 @@ protected void createProducer(Map<String, ?> config, String producerClientId) {
    * @return Desired replication factor of sample store topics, or {@code null} if failed to resolve replication factor.
    */
   protected short sampleStoreTopicReplicationFactor(Map<String, ?> config, AdminClient adminClient) {
-    if (_sampleStoreTopicReplicationFactor == null) {
-      short numberOfBrokersInCluster;
+    if (_sampleStoreTopicReplicationFactor != null) {
+      return _sampleStoreTopicReplicationFactor;
+    }
+
+    int maxRetryCount = Integer.parseInt(config.get(MonitorConfig.FETCH_METRIC_SAMPLES_RETRY_COUNT_CONFIG).toString());
+    long waitingTime = Long.parseLong(config.get(MonitorConfig.FETCH_METRIC_SAMPLES_WAITING_MS_CONFIG).toString());
+    int retryCount = 0;
+
+    short numberOfBrokersInCluster = 0;
+    while (retryCount <= maxRetryCount) {
       try {
-        numberOfBrokersInCluster = (short) adminClient.describeCluster().nodes().get(CLIENT_REQUEST_TIMEOUT_MS,
-                                                                                     TimeUnit.MILLISECONDS).size();
-      } catch (InterruptedException | ExecutionException | TimeoutException e) {
-        throw new IllegalStateException("Auto creation of sample store topics failed due to failure to describe cluster.", e);
-      }
-      if (numberOfBrokersInCluster <= 1) {
-        throw new IllegalStateException(String.format("Kafka cluster has less than 2 brokers (brokers in cluster=%d, zookeeper.connect=%s)",
-                                                      numberOfBrokersInCluster, config.get(ExecutorConfig.ZOOKEEPER_CONNECT_CONFIG)));
+        try {
+          numberOfBrokersInCluster = (short) adminClient.describeCluster().nodes().get(CLIENT_REQUEST_TIMEOUT_MS,
+                  TimeUnit.MILLISECONDS).size();
+        } catch (InterruptedException | ExecutionException | TimeoutException e) {
+          throw new IllegalStateException("Auto creation of sample store topics failed due to failure to describe cluster.", e);
+        }
+
+        if (numberOfBrokersInCluster <= 1) {
+          throw new IllegalStateException(String.format("Kafka cluster has less than 2 brokers (brokers in cluster=%d, zookeeper.connect=%s)",
+                  numberOfBrokersInCluster, config.get(ExecutorConfig.ZOOKEEPER_CONNECT_CONFIG)));
+        }
+
+        return (short) Math.min(DEFAULT_SAMPLE_STORE_TOPIC_REPLICATION_FACTOR, numberOfBrokersInCluster);
+      } catch (IllegalStateException e) {
+        if (retryCount == maxRetryCount) {
+          throw e;
+        }
```

Comment:
This retry logic keeps retrying if `numberOfBrokersInCluster <= 1`. What is the intention of retry if the cluster has less than or equal to 1 broker? Do we expect the number of alive nodes in the cluster to increase between retries?

Reply:
Yes, it may happen (rarely) with Kafka, that brokers initialization is too slow. If the CC started too and this line of code is reached during the init of the Kafka then this exception could happen. This failure can prevent with this retry mechanizm since further couple of seconds/mins could be enough to the Kafka brokers to be available.